src/util/route.js

```diff
@@ -6,13 +6,13 @@ export function isSameRoute (a: Route, b: ?Route): boolean {
   } else if (a.path && b.path) {
     return (
       a.path === b.path &&
-      a.hash === b.hash &&
+      (a.hash || '') === (b.hash || '') &&
```

Comment:
That's weird, this line: https://github.com/vuejs/vue-router/blob/next/src/util/route.js#L3 is telling that `b` can be a route or `null` so giving it a location should indeed raise an error...